Query Mbed TLS configuration from ssl_client2 and ssl_server2

andresag01 opened the pull request:
## Description
This PR:
* fixes #2029
* fixes #2030

The problem was that the detection of the current configuration in ssl-opt.sh used to be based on a fragile combination of calls to grep and sed which did not work in some cases and is not very portable. This PR replaces that mechanism by extending the command line interface for ssl_client2 and ssl_server2 to allow querying the current Mbed TLS configuration. For example, to check if MBEDLTS_AES_C was enabled at compile time, we simply need to run the command:
```
programs/ssl/ssl_server2 query_config=MBEDTLS_AES_C
```
A return value of 0 means that the feature is enabled and 1 means it was not. If the macro has a non empty expansion, this will be printed to stdout.

## Status
**READY**

## Requires Backporting
[edited by mpg] Backporting required to 2.16 as this fixes issues in the tests - #2389
No need to backport for 2.7 as the issues were only introduced later (2.13).

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

# Additional comments
* This PR adds a dependency of ssl_client2 and ssl_server2 on the file query_config.c, so I am not sure whether the visualc files also need to be updated. However, I do not have a system with visual studio to test this.
* I tested this on OS X and there are 4 ssl-opt.sh tests failing on my system, but the failures are also present in development.

andresag01 commented on scripts/config.pl:
```diff
@@ -46,7 +46,7 @@
 $0 [-f <file> | --file <file>] [-o | --force]
                    [set <symbol> <value> | unset <symbol> | get <symbol> |
-                        full | realfull | baremetal]
+                        full | realfull | baremetal | all]
```
Adding this option was necessary to automate the generation of the query_config.c file because the previous options did not uncomment the module defines in the config.h

hanno-arm commented on scripts/config.pl:
```diff
@@ -218,8 +220,10 @@
+    if ($action eq "full" || $action eq "realfull" || $action eq "all" || $action eq "baremetal" ) {
+        if ($action ne "all" && $line =~ /name SECTION: Module configuration options/) {
+            $done = 1;
+        } elsif ($line =~ /name SECTION: Customisation configuration options/) {
```
Maybe I'm blind, but how is the new cold functionally different from the old?

hanno-arm commented on scripts/generate_query_config.pl:
```diff
@@ -0,0 +1,79 @@
+    }
+}
+
+# Read the fill format file into a string
```
Typo: `fill` -> `full`

hanno-arm commented:
@andresag01 What was the reason not to add a separate example program for querying the configuration? I find it unintuitive that's done via an additional command line parameter of `ssl_server2`.

andresag01 commented:
@hanno-arm: Thanks for the review.

>Apart from that, it's not clear to me why the configuration querying was introduced as part of ssl_client2 and ssl_server2, and not as a separate program in the programs/test folder. 

I personally do not mind whether it is a separate program or part of ssl_client2 or ssl_server2. This is what @sbutcher-arm suggested and I don't have any objections to either approach. As you can see, the changes for ssl_server2 and ssl_client2 are very small and the query code is fully contained in the `query_config.c` file. Adding a new app would take slightly more code, but is also very straight forward.

I also replied to the other comment in your review. I will wait for a response before any rework.

andresag01 commented:
@hanno-arm: I have reworked the PR to remove the changes to config.pl and to not use the C preprocessor.

k-stachowiak commented on programs/ssl/ssl_client2.c:
```diff
@@ -358,6 +362,8 @@ struct options
     int etm;                    /* negotiate encrypt then mac?              */
 } opt;
 
+int query_config( const char *config );
```
I assume that this line has been put here and in the other program in order not to introduce a header file, which may mess with the concept of the public API. Nevertheless, this looks a bit odd, and I would opt for a cleaner solution. I think that @hanno-arm's suggestion to put the code in a separate program would solve the issue in a simpler way. I would also think, how this change impacts the value of the program as an example - maybe we should avoid addind non-critical complexity to the example programs.

andresag01 replied:
This is here because I did not think it was necessary to add a header file because of a 1 line function declaration. I am not particularly worried about the API because this is an example application, not the library.

>I think that @hanno-arm's suggestion to put the code in a separate program would solve the issue in a simpler way.

Please see the discussion with @sbutcher-arm and @hanno-arm regarding stand-alone app vs ssl_client2/ssl_server2 extension.

k-stachowiak commented on tests/ssl-opt.sh:
```diff
@@ -165,22 +165,34 @@ requires_config_disabled() {
+    #
+    # Note that if the configuration is not defined or is defined to nothing,
+    # the output of this function will be an empty string.
+    ${P_SRV} "query_config=${1}"
```
If only the server is used for the query_config functionality, why was it also added to the client example?

andresag01 replied:
Please see the discussion of stand-alone app vs ssl_client2/ssl_server2.

hanno-arm commented:
> This is my preference. The idea being, if presented with ssl_client2 and ssl_server2 as binaries, you can query their build configurations if you're using them. If we have a separate application, it has to be with those binaries to be usable.

@sbutcher-arm I see your point, but for use of Mbed TLS outside of `ssl_server2/ssl_client2` I think it's unclear why you would use server/client example applications to query the configuration.

I am quite strongly in favour of a separate application for configuration querying (which doesn't mean we can't _also_ have it as a command line parameter in `ssl_server2/ssl_client2`), but it's not a blocker for me, and in the end it's @sbutcher-arm's decision anyway.

sbutcher-arm commented:
>  I think it's unclear why you would use server/client example applications to query the configuration.

Because you're presented with a binary, you don't know who built it or its origins, and you want to know how it was configured.

hanno-arm commented:
> Because you're presented with a binary, you don't know who built it or its origins, and you want to know how it was configured.

Yes, but I meant _outside_ of the use through `ssl_server2` and `ssl_client2`.

andresag01 commented:
@hanno-arm @sbutcher-arm: As I mentioned before, I personally dont have a strong feeling about either approach. How about if we have the option to query the configuration through `ssl_client2/ssl_server2` AND as a separate `query_config` application? Its not hard to do this and most of the code is going to be in a shared file anyways...

hanno-arm commented:
@andresag01 I'd like that.

sbutcher-arm commented:
@andresag01 - I'm ok with that approach, as long as we can do it quickly.

andresag01 commented:
@sbutcher-arm @hanno-arm @k-stachowiak: I have created the stand-alone app (which I called query_compile_time_config to not collide with the name of the shared file query_config.c), hopefully this addresses the concerns raised.

Could you please review once again?

sbutcher-arm commented on programs/ssl/query_config.c:
```diff
@@ -0,0 +1,2465 @@
+#define mbedtls_printf printf
+#endif /* MBEDTLS_PLATFORM_C */
+
+/* Include all the headers with public APIs in case they modify any configs */
```
The headers don't (or shouldn't) 'modify any configs' but define defaults if they're not already defined. They shouldn't replace or redefine them.

Can we make the comment clearer?

sbutcher-arm commented on programs/ssl/query_config.c:
```diff
@@ -0,0 +1,2465 @@
+#define MACRO_EXPANSION_TO_STR(macro)   MACRO_NAME_TO_STR(macro)
+#define MACRO_NAME_TO_STR(macro)        strlen( #macro ) > 0 ? #macro "\n" : ""
+
+int query_config( const char *config )
```
This function looks like a lot of work. When we discussed adding this feature, I imagined we'd add one or two symbols to fix the current problems and not the entire set, so I wasn't expecting this function and the whole set!

I'm fine with what we have, but it does feel like a little overkill for the bug we had!

andresag01 replied:
Well, my thinking was that it would be slightly irritating to have to modify this program manually whenever we want to add a new config or new tests...

sbutcher-arm replied:
You turned it around so quickly, the time spent on it doesn't matter, I guess  you're just overachieving. ;)

sbutcher-arm commented on programs/ssl/query_config.c:
```diff
@@ -0,0 +1,2465 @@
+    }
+#endif /* MBEDTLS_HMAC_DRBG_RESEED_INTERVAL */
+
+#if defined(MBEDTLS_HMAC_
```
It appears we display nothing for configurations not compiled in, but we do return a return value of 1, which I think is the correct behaviour.

Can we say as much in a comment below?
`/* If the symbol is not found, return an error */`

andresag01 replied:
I can certainly put in a comment somewhere in the function.

sbutcher-arm commented:
Travis is failing! Two issues:
```
0.04s$ tests/scripts/check-generated-files.sh
'programs/ssl/query_config.c' was either modified or deleted by 'scripts/generate_query_config.pl'
The command "tests/scripts/check-generated-files.sh" exited with 1.
```
That's acceptable I guess as `programs/ssl/query_config.c` wasn't previously present.

This however looks like a problem:
```
/home/travis/build/ARMmbed/mbedtls/programs/ssl/query_config.c: In function ‘query_config’:
/home/travis/build/ARMmbed/mbedtls/programs/ssl/query_config.c:115:9: error: zero-length gnu_printf format string [-Werror=format-zero-length]
         mbedtls_printf( MACRO_EXPANSION_TO_STR( MBEDTLS_HAVE_ASM ) );
         ^
/home/travis/build/ARMmbed/mbedtls/programs/ssl/query_config.c:147:9: error: zero-length gnu_printf format string [-Werror=format-zero-length]
         mbedtls_printf( MACRO_EXPANSION_TO_STR( MBEDTLS_HAVE_TIME ) );
         ^
/home/travis/build/ARMmbed/mbedtls/programs/ssl/query_config.c:155:9: error: zero-length gnu_printf format string [-Werror=format-zero-length]
         mbedtls_printf( MACRO_EXPANSION_TO_STR( MBEDTLS_HAVE_TIME_DATE ) );
         ^
```

andresag01 commented:
@sbutcher-arm: Thanks for reviewing. With regards to the failures:
* I will look into the check-generated-files.pl failure.
* The second failure is because apparently GCC does not like a print statement with an empty forma string i.e. `mbedtls_printf( "" )`. I will look into it, probably something like `mbedtls_printf( "%s", "" )` fixes the problem.

andresag01 commented on programs/test/query_compile_time_config.c:
```diff
@@ -0,0 +1,56 @@
+        return( MBEDTLS_EXIT_FAILURE );
+    }
+
+    return( query_config( argv[1] ) );
```
I can do that, but the reason why I originally didnt is because that value is potentially different across operating systems/platforms. I wanted to have a well defined value that never changes, so I chose 0 and 1

andresag01 commented:
@sbutcher-arm: I believe I have now addressed your comments.

@k-stachowiak: Thanks fore reporting the problems with Visual Studio. Unfortunately, I do not have access to the toolchain to test this myself, so could you please confirm that the following patch fixes the inclusion problem of query_config.c?
```
diff --git a/visualc/VS2010/ssl_client2.vcxproj b/visualc/VS2010/ssl_client2.vcxproj
index 1d44fa783..a960facf0 100644
--- a/visualc/VS2010/ssl_client2.vcxproj
+++ b/visualc/VS2010/ssl_client2.vcxproj
@@ -20,6 +20,7 @@
   </ItemGroup>
   <ItemGroup>
     <ClCompile Include="..\..\programs\ssl\ssl_client2.c" />
+    <ClCompile Include="..\..\programs\ssl\query_config.c" />^M
   </ItemGroup>
   <ItemGroup>
     <ProjectReference Include="mbedTLS.vcxproj">
diff --git a/visualc/VS2010/ssl_server2.vcxproj b/visualc/VS2010/ssl_server2.vcxproj
index d06e0628e..06a91cb49 100644
--- a/visualc/VS2010/ssl_server2.vcxproj
+++ b/visualc/VS2010/ssl_server2.vcxproj
@@ -20,6 +20,7 @@
   </ItemGroup>
   <ItemGroup>
     <ClCompile Include="..\..\programs\ssl\ssl_server2.c" />
+    <ClCompile Include="..\..\programs\ssl\query_config.c" />^M
   </ItemGroup>
   <ItemGroup>
     <ProjectReference Include="mbedTLS.vcxproj">
```
(Note the ^M character at the end of every line...)

Also, I did not quite understand how you solved the macro expansion problem. Did you mean that the following patch fixes the problem?
```
diff --git a/programs/ssl/query_config.c b/programs/ssl/query_config.c
index 8ef5d3739..5a34ee781 100644
--- a/programs/ssl/query_config.c
+++ b/programs/ssl/query_config.c
@@ -107,8 +107,7 @@
 #include <string.h>

 /* Helper macros to convert a macro or its expansion into a string */
-#define MACRO_EXPANSION_TO_STR(macro)   MACRO_NAME_TO_STR(macro)
-#define MACRO_NAME_TO_STR(macro)        strlen( #macro ) > 0 ? #macro "\n" : ""
+#define MACRO_EXPANSION_TO_STR(macro) strlen( #macro ) > 0 ? #macro "\n" : ""

 int query_config( const char *config )
 {
```
Does this still print out the correct result to stdout? For example, does this give:
```
$ ./programs/ssl/ssl_client2 query_config=MBEDTLS_PLATFORM_CALLOC_MACRO
calloc
```
and a return value of 0?

k-stachowiak commented:
@andresag01 

Regarding the 1st issue, I confirm, that the way you added the `query_config.c` file to the project files gets the projects to build (given the 2nd issue has been fixed too). But it has two issues:

* This is also necessary in the `query_compile_time_config` project
* Unfortunately, It won't work in a general case, as the project files are generated by the `scripts/generate_visualc_files.pl`, which uses the `scritps/data_files/vs2010-app-template.vcxproj` for generating program. The template assumes, that each program only consists of a single source file. The only quick fix that comes to my mind at the moment is something along the lines of:
```
#if MSVC
#    include "query_config.c"
#else
int query_config( const char *config );
#endif
```
It is a bit dirty, but with some tweaks (e.g. changing `query_config.c` to a header and making `query_config()` inline), may give a relatively painless solution.

Regarding the 2nd issue, the problem is, that Visual C++ (tested on VS 2017) seems to behave in a non-confirming way. What should work is something like:
```
#if MSVC
#    define MACRO_EXPANSION_TO_STR(macro)   strlen( #macro ) > 0 ? #macro "\n" : ""
#else
#    define MACRO_EXPANSION_TO_STR(macro)   MACRO_NAME_TO_STR(macro)
#    define MACRO_NAME_TO_STR(macro)        strlen( #macro ) > 0 ? #macro "\n" : ""
#endif
```
However I don't have enough knowledge to tell you the exact documents that define the behavior. A <1h search didn't lead me to a clear explanation of this, so my advice is purely empyrical, which may not be good enough for production. Unfortunately, I do not have more resources to investigate the documentation right now, but I am ready to test any solutions that come to your mind on my setup.

andresag01 commented:
@k-stachowiak: Thanks for the info. I have reworked according to what you said. Could you please review and test once again?

k-stachowiak commented:
@andresag01 The project setup changes look good. I wouldn't call it perfect, as it may be confusing for some to figure out, how it works, but I find it good enough for its purpose. Also if the scripts are to be reworked in the future, it will be a better context for finding the best design for this aspect.

Regarding the macro expansion, unfortunately, while it builds, it doesn't lead to a proper expansion of the macros' values (as you predicted). It only prints the macro name or nothing. The exit value, however, is correct.

andresag01 commented:
I have fixed and tested the application to work with Visual Studio. @k-stachowiak please review and test once again.

sbutcher-arm commented:
retest

mpg commented:
The lastest commit fails to build on Jenkins Linux builders in `tests-ref-configs.pl` in `config-suite-b`:
```
[ 35%] Building C object programs/ssl/CMakeFiles/ssl_client2.dir/query_config.c.o

/var/lib/build/programs/ssl/query_config.c:2419:33: error: too many arguments provided to function-like macro invocation

        MACRO_EXPANSION_TO_STR( MBEDTLS_SSL_CIPHERSUITES );

                                ^

/var/lib/build/include/mbedtls/config.h:105:5: note: expanded from macro 'MBEDTLS_SSL_CIPHERSUITES'

    MBEDTLS_TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256

    ^

/var/lib/build/include/mbedtls/ssl_ciphersuites.h:153:58: note: expanded from macro 'MBEDTLS_TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256'

#define MBEDTLS_TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256  0xC02B /**< TLS 1.2 */

                                                         ^

/var/lib/build/programs/ssl/query_config.c:115:9: note: macro 'MACRO_NAME_TO_STR' defined here

#define MACRO_NAME_TO_STR(macro)                                        \

        ^

/var/lib/build/programs/ssl/query_config.c:2419:9: error: use of undeclared identifier 'MACRO_NAME_TO_STR'

        MACRO_EXPANSION_TO_STR( MBEDTLS_SSL_CIPHERSUITES );

        ^

/var/lib/build/programs/ssl/query_config.c:114:41: note: expanded from macro 'MACRO_EXPANSION_TO_STR'

#define MACRO_EXPANSION_TO_STR(macro)   MACRO_NAME_TO_STR(macro)

                                        ^

2 errors generated.
```

andresag01 commented:
@mpg: Thanks for reviewing. You are right regarding the problem with macros defined to a comma separated list of items. I tried for a bit to look for a solution to the problem, but could not come up with a simple one only using the C preprocessor. So I decided to extend the script to simply not include the problematic macro in the query_config.c autogenerated code. While unfortunate I do not think this will be a problem in practice because the build system does not check that macro for any of the tests (as far as I know).

andresag01 commented:
@mpg: Thanks fore reviewing and apologies for letting these issues slip through. I rebased the PR after testing and did not realise that the development has new macros which needed to be included in query_config.c. I have generated the file again, but had to exclude the function-like macro MBEDTLS_CHECK_PARAMS in the same way as SSL_CIPHERSUITES.

> avoid using non-Unix line endings in ./scripts/generate_visualc_files.pl on lines 96 and 99
Sorry about this. I have replaced the carriage return character with a \r as this is how the newlines in the generated visualc files are used.

Waiting for the CI to finish, hopefully there will not be more issues.

andresag01 commented:
I think only the timing tests failed in CI

andresag01 commented:
retest

andresag01 commented:
@sbutcher-arm @mpg: It looks like there is a Windows failure (more like a warning really) when compiling the query_config.c file. This is because Visual Studio does not like when a function-like macro receives an "empty" arguments. For example, something like this gives a warning in visual studio:
```
#define DO_SOMETHING(X) #X

DO_SOMETHING(MBEDTLS_AES_C) // This line gives a warnings because MBEDTLS_AES_C is defined to nothing
```

The problem does not happen in GCC and Clang though. Do you think its acceptable to just disable that particular warning in visual studio for the problematic query_config.c file? Any other ideas?

mpg commented:
@andresag01 I think if MSVC has a pragma directive to disable that particular warning, then doing so in that file is my favourite solution to the issue. Clearly in that case we're fully aware of what we're doing and we don't care about this warning.

sbutcher-arm commented:
I think it's acceptable to have a `#pragma` in `query_config.c` given it's just a tool and not part of the library.

andresag01 commented:
retest

andresag01 commented:
@sbutcher-arm @mpg: I have added code to suppress the Visual studio warnings in the problematic piece of code only. Waiting for the CI to finish

mpg commented:
@k-stachowiak You reviewed an earlier version of this PR, which has been reworked since then. Can you please review the PR again so that we can hopefully merge it soon?

mpg commented:
@andresag01 Please don't use the "retest" comment, it triggers the old CI job (that tends to cause issues with the CI). The new CI job can be re-triggered by clicking the "details" link in github then the "rerun" button in the top toolbar of the resulting page (just right of "artifacts").

mpg commented:
@andresag01 `tests/scripts/check-generated-files.sh` is still failing on the CI (and locally on my machine). Can you please look into it?

andresag01 commented:
@mpg: Thanks for pointing out the check-generated-files issue. I pushed a fix for this.

andresag01 commented:
@mpg: Thanks for approving.

@sbutcher-arm: Do you think its worth backporting any of these changes?

mpg commented:
@andresag01 I know the question was for Simon, but IMO if we want to be able to use this in test scripts, we'll need to backport it.

sbutcher-arm commented:
@hanno-arm / @k-stachowiak - Could one of you please re-review and approve this PR? We need this PR to fix some CI problems which are limiting our test coverage.

andresag01 commented:
@mpg: I just realised that the failing tests that these changes address are not in Mbed TLS 2.7, so I do not think it needs not be backported to that version. I will backport to Mbed TLS 2.16 (our new LTS)

andresag01 commented:
@mpg: You are right, the only conflict when merging to 2.16 is the ChangeLog. But I created the backport already: https://github.com/ARMmbed/mbedtls/pull/2389

mpg commented:
@andresag01 This PR now has conflicts in `ssl-opt.sh`. Please resolve them.

andresag01 commented:
@mpg: I rebased to resolve the conflicts.

mpg commented:
@andresag01 The CI wants you to re-rerun `scripts/generate_query_config.pl` :)

mpg commented:
@andresag01 Once you've updated the generated file, can you please:
- ping @k-stachowiak and myself for re-review
- check if the PR applies cleanly to 2.16 as well: if so, mention it in the description, else create a backport PR?
Thanks!

andresag01 commented:
@mpg @k-stachowiak: I fixed the check-generated-files.pl problem. Also, the Mbed TLS 2.16 PR is at https://github.com/ARMmbed/mbedtls/pull/2389

mpg commented:
2.16 backport is now ready for merge in #2389 - no other backport required, so removing "needs: backports"

sbutcher-arm commented:
Reviewed, approved, passing CI, and backports ready, so promoting to 'ready for merge'.

Patater commented:
This fails `./tests/scripts/check-generated-files.sh`

``` 'programs/ssl/query_config.c' was either modified or deleted by 'scripts/generate_query_config.pl' ```

Demoting to needs work.

Patater commented:
I have a fix available that can be force pushed to this branch. It required rebasing to pick up PSA-related changes so those could be added to `query_config`, so a diff will be a bit more difficult.

Available here: https://github.com/Patater/mbedtls/tree/iotssl-2555-config-detection-2

andresag01 commented:
@Patater: Agreed, the problem is that the branch that this PR was applied to contains more config options that were not in the version of the library that this was not based on. Would you like me to apply your patch to this PR? Note that this would cause the CI on this PR to fail. I think it would be best if the gatekeeper runs the generate files script after merging.

Patater commented:
Applying my patch wouldn't work unless the base was also updated. I'd prefer to rebase this PR and include the updated generated files, so we can review that code. But, since this is auto-generated code, we can do it as part of gatekeeping.

andresag01 commented:
@Patater: Thanks for clarifying.
